### PR TITLE
[azd down] ux updates

### PR DESCRIPTION
**Pull request by vhvb1989**
based on: https://www.figma.com/proto/iIRRiGPuKIfaSFfMJXcYle/azd-resource-provisioning?page-id=1434%3A135002&node-id=1436%3A135020&viewport=5845%2C-785%2C1&scaling=min-zoom&starting-point-node-id=1436%3A135020

fix: https://github.com/Azure/azure-dev/issues/1434
fix: https://github.com/Azure/azure-dev/issues/1588

### CODE REVIEW

File: cli/azd/cmd/down.go [hunk 1]
```diff
@@ -97,30 +98,34 @@ func newDownAction(
...
+	// silent manager for running Plan()
+	infraManager, err := createProvisioningManager(ctx, a, &project.MutedConsole{ParentConsole: a.console})
 	if err != nil {
 		return nil, fmt.Errorf("creating provisioning manager: %w", err)
 	}
 
+	// Command title
+	a.console.MessageUxItem(ctx, &ux.MessageTitle{
+		Title:     "Deleting all resources and deployed code on Azure (azd down)",
+		TitleNote: "Local application code is not deleted when running 'azd down'.",
```
Comment (weikanglim) [inline cli/azd/cmd/down.go:110]: @savannahostrowski @Austinauth Do you think we could make these more concise for better UX?

I was thinking along the lines of:

"Delete Azure resources (azd down)"
"Cleaning up provisioned infrastructure and services in Azure. Local application files will be kept."

I'm still not sure about the mention about "application files" being kept. Most computer users may expect programs to not remove their files unless explicitly prompted so (products that violate this probably don't end up living long), and maybe the help text/description suffices. It feels close to "Windows updates" reassurance of "your files are exactly where you kept them", which always reassures me less as a user that Windows needs to bring that up (and because it has happened, and still happens today).

File: cli/azd/cmd/down.go [hunk 2]
```diff
@@ -137,7 +142,29 @@ func (a *downAction) Run(ctx context.Context) (*actions.ActionResult, error) {
 		return nil, fmt.Errorf("saving environment: %w", err)
 	}
 
-	return nil, nil
+	return &actions.ActionResult{
+		Message: &actions.ResultMessage{
+			Header: "Your Azure resources have been destroyed.",
```
Comment (weikanglim) [inline cli/azd/cmd/down.go:147]: @savannahostrowski @Austinauth Does "destroyed" bring in a different meaning than "deleted"? It seems to suggest stronger wording than expected?

File: cli/azd/cmd/down.go [hunk 3]
```diff
@@ -97,30 +98,34 @@ func newDownAction(
...
-		a.console,
-		a.commandRunner,
-		a.accountManager,
-		a.userProfileService,
-		a.subResolver,
-		a.alphaFeatureManager,
-	)
-
+	// silent manager for running Plan()
+	infraManager, err := createProvisioningManager(ctx, a, &project.MutedConsole{ParentConsole: a.console})
```
Comment (weikanglim) [inline cli/azd/cmd/down.go:102]: If we're surfacing `MutedConsole`, then it probably should move to `output` package instead of `project`. Can be a separate PR.

File: cli/azd/pkg/input/console.go [hunk 4]
```diff
@@ -229,6 +229,7 @@ func (c *AskerConsole) StopSpinner(ctx context.Context, lastMessage string, form
 
 	c.spinner.StopMessage(lastMessage)
 	_ = c.spinner.Stop()
+	c.currentIndent = ""
```
Comment (weikanglim) [inline cli/azd/pkg/input/console.go:232]: Is there a more concise rule on how "currentIndent" is handled by the `console`, that now can be documented on the member variable `console.currentIndent`?

Comment (vhvb1989) [reply cli/azd/pkg/input/console.go:232]: no. This was actually a bug. 
indentation should be reset whenever we stop a spinner.  Or it should come back to previous state, in case we support multi-levels of indentation.
But we are not yet in a multi-level (more than 2), so for now this is fine.

File: cli/azd/cmd/down.go (hunk 1 above)
Comment (Austinauth) [reply cli/azd/cmd/down.go:110]: @weikanglim, Savannah and I discussed this previously when we made the change.

The reason for this title and description is because we observed that some users failed to understand where the code was deleted vs were it was kept (including myself on first use).

I am okay consolidating the primary title back to "Deleting Azure resources (azd down)", however this puts more pressure on the title note to be more descriptive .The proposed title not change does not feel any more clear to me than what was initially proposed.

File: cli/azd/cmd/down.go (hunk 2 above)
Comment (Austinauth) [reply cli/azd/cmd/down.go:147]: Agree, we should probably just use delete here.

Comment (savannahostrowski) [reply cli/azd/cmd/down.go:147]: Yep - +1!

**Comment by Austinauth** [comment]
> @vhvb1989 there was ask around printing the portal link - #1924 before confirmation. So users can take a look at what they are deleting. Can you also include here or can it be separate?
> 
> cc: @savannahostrowski @Austinauth

@rajeshkamal5050, I'm okay with providing the portal link before they confirm deletion. I'm not sure I understand the question though. What do you mean when you say " Can you also include here or can it be separate"?

**Comment by rajeshkamal5050** [comment]
> was that directed at Victor?

Yes :-)

File: cli/azd/pkg/output/ux/list_as_text.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+// generates text from a list of strings, for example:
+// ["foo"]        					=> "foo"
+// ["foo", "bar"] 					=> "foo and bar"
+// ["foo", "bar", "axe"] 			=> "foo, bar and axe"
+// ["foo", "bar", ..., ..., "axe"] 	=> "foo, bar, ..., ... and axe"
+func ItemsAsText(items []string) string {
+	count := len(items)
+	if count < 1 {
+		log.Panic("calling itemsCountAsText() with empty list.")
+	}
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/output/ux/list_as_text.go:21]: errors are for customers/folks consuming azd.
panics are for developers coding azd.  

This `panic` is a message for a developer which should be seen during testing azd and which should be fixed before it is released to the public.

When a customer gets a `panic` message, it means a development escape. And the app needs urgent attention.

When a customer gets an `error`, it means the operation went wrong but it is not due to problematic source code. For example, error responses from Azure, missing files, bad user configuration, etc.  Where azd is doing its job good, but something else is causing an error.

A panic means azd is doing something wrong. In this specific example, developers must ensure that when they call this method, there are items in the `[]string` and do not use this method if there is a remote chance for the array to be empty.

**Comment by vhvb1989** [comment]
@rajeshkamal5050  @savannahostrowski  @Austinauth , I added the link to resource group as part of the confirmation as:

image

This is for one resource group.

**Comment by vhvb1989** [comment]
More than one ResourceGroup view:

image

**Comment by rajeshkamal5050** [comment]
> More than one ResourceGroup view:
> 
> image

Thanks Victor. Since you are adding links. Can you also make sure of this issue - https://github.com/Azure/azure-dev/issues/1588
- Show RG portal link for success/failure scenarios of deploy and provision.
- Show Correlation ID for success/failure scenarios. Ask from CSS.

**Comment by vhvb1989** [comment]
Adding correlation id for the deployment:

`azd provision`
image

`azd up`
image


The correlationId is only displayed after `provision` operation has been executed. Either from running `azd provision` alone or from running `azd up`.

It won't be displayed when running `azd deploy`:

image

File: cli/azd/cmd/actions/action.go [hunk 6]
```diff
@@ -17,8 +17,9 @@ func (a ActionFunc) Run(ctx context.Context) (*ActionResult, error) {
 
 // Define a message as the completion of an Action.
 type ResultMessage struct {
-	Header   string
-	FollowUp string
+	Header        string
+	CorrelationId string
```
Comment (weikanglim) [inline cli/azd/cmd/actions/action.go:21]: This seems wrong...? Should a `CorrelationId` be part of an ActionResult?

File: cli/azd/pkg/tools/azcli/azcli.go [hunk 7]
```diff
@@ -347,3 +356,25 @@ func clientOptionsBuilder(httpClient httputil.HttpClient, userAgent string) *azs
 		WithTransport(httpClient).
 		WithPerCallPolicy(azsdk.NewUserAgentPolicy(userAgent))
 }
+
+type PurgeImplementation func(ctx context.Context, subscriptionId string, name string, location string) error
+
+func (cli *azCli) getPurgeImplementation(resourceType PurgeResourceType) PurgeImplementation {
```
Comment (weikanglim) [inline cli/azd/pkg/tools/azcli/azcli.go:362]: Does this make it harder for us to refactor away from the uber `azCli` struct in favor of individual clients?

File: cli/azd/cmd/actions/action.go (hunk 6 above)
Comment (weikanglim) [reply cli/azd/cmd/actions/action.go:21]: I don't think we should be outputting a "provision" correlation ID in the "up" flow. It seems hard for a user to guess that the correlation ID belonged to the infrastructure provisioning operation out of all the things that happened. 

I'd encourage for a design in `provision` that would support logging a message that easily understood to be part of the operation.

File: cli/azd/cmd/up.go [hunk 8]
```diff
@@ -146,6 +146,8 @@ func (u *upAction) Run(ctx context.Context) (*actions.ActionResult, error) {
 	if err != nil {
 		return nil, err
 	}
+	// Use the correlation id from the provision operation
+	deployResult.Message.CorrelationId = provisionResult.Message.CorrelationId
```
Comment (ellismg) [inline cli/azd/cmd/up.go:150]: This feels wrong to me at some level - the deploy operation is going to do a bunch of work (deploying each service) and each piece of work logically has its own correlation ID.  I think long term we would want to aggregate all the correlation ids generated by each step of the `up` flow and display them all.

I suspect that this is the path of least resistance in order for us to display the correlation id from the provision operation at the end of the `up` flow - so I'm not going to block on this feedback, but I think we need to think long term about how we want to handle this sort of thing in the up flow.

Honestly, I'm not sold on the value of printing these correlation IDs to the end user on all these operations.  My mental model is that these are not useful except in the cases where I want to open a support ticket.  I could see printing them on error paths (as part of context of the error), and perhaps `log.Print`'ing them in all cases (since if you are running with `--debug, they could be useful), but I do feel like the UX of the tool got a little worse with this change when viewed through the lens of someone who is new to Azure.

In practice, it seems like you can [recover the correlation id from the deployment itself](https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-history?tabs=azure-portal) (which is retained), so it's not as if dropping this value on the floor makes it unrecoverable.

File: cli/azd/pkg/tools/azcli/azcli.go (hunk 7 above)
Comment (ellismg) [reply cli/azd/pkg/tools/azcli/azcli.go:362]: I'm also not sold on this part of the refactoring - Looking at the callers of this method in `bicep_provider.go` we already have context when we call `PurgeResource` as to which type of resource we want to purge (and it's further enforced by the new `PurgeResource` API you added, since I have to pass in a type) and so it feels to me like we should continue to just have the `AzCli` type expose the resource specific primitives and have the code that decides which of these to call live in the bicep layer, which already has a bunch of other policy decisions.  It also feels like long term the set of information we need to purge a resource will be specific to that type, and so trying to group everything under a single PurgeResource operation will be confusing.

I'm sure there are a bunch of pros off this design, but I'm personally struggling to see them, could you help me understand what pushed you towards this refactor?

**Review by ellismg — COMMENTED** [review]
Really excited to pull this down and play around with the new UX.  I have stars in my eyes looking at some of the screenshots you shared.

Some questions to help be better understand what's going on and a few comments about the correlation ID stuff (I know this was an ask from the support folks, but I'm a little worried that just throwing correlation IDs at the user is going to make the tool feel a little less friendly, so I'd love if we could be a little mindful about how we introduce them).

File: cli/azd/cmd/actions/action.go (hunk 6 above)
Comment (vhvb1989) [reply cli/azd/cmd/actions/action.go:21]: I will remove the correlationId from the changes.
It was requested by Rajesh!!! hahah,, and he did not hear my reasons about why it was not good to add it until you and Matt say it as well =D

File: cli/azd/pkg/tools/azcli/azcli.go (hunk 7 above)
Comment (vhvb1989) [reply cli/azd/pkg/tools/azcli/azcli.go:362]: > I'm sure there are a bunch of pros off this design

Not really!! hahaha... I was just hoping to reduce the `purge` api from the `az` interface to just one. And use that one to call the specific implementation.  This is because I know we will be adding purge support for some more stuff in the future and noticed they all use the same input, just for different clients.

I will revert this, I don't have a strong need to keep it

File: cli/azd/cmd/up.go (hunk 8 above)
Comment (vhvb1989) [reply cli/azd/cmd/up.go:150]: I didn't want to have the correlatioID on happy paths as well.  For example, on success cases.
Rajesh said that sometimes, even there is a success message, things might be working, and then we need that Id for creating the ticket.

I am just happy that you and Wei show up here and we'll be removing this correlationId from happy paths :)

Comment (rajeshkamal5050) [reply cli/azd/cmd/up.go:150]: @ellismg @vhvb1989 
CSS was asking for Azd to support/enable log files. So by default, we have debug logs with rotation policies in place.

Was thinking there would one only one correlation-if for our deployment. If more than one, should we instead log correlation id's to logs?

Comment (ellismg) [reply cli/azd/cmd/up.go]: > CSS was asking for Azd to also support/enable log files. So by default, we have debug logs with rotation policies in place. Thought there would one only one correlation-if for our deployment. If more than one, should we instead log correlation id's to logs?

After reading a little more, I think what we should be doing is something like this (to be clear, I'm not suggesting we do this as part of this PR)

1. When `azd` starts up, we should generate a random UUID, and add it to our context (we should also `debug.Log` it)
2. We should make the necessary changes to our clients such that we set an `x-ms-correlation-request-id` header on all the requests we make (this allows multiple independent requests to be correlated back to a single run of the CLI).
3. We should update our logging policy to add this header to the list of allowed headers to log (again, we want to do this globally, for all clients)

I think with these changes, we'll be logging both the client specific request ID (which is generated per operation) as well as an ID which can be used to correlate multiple REST calls across a single AZD operation.

This should give us the complete set of IDs that CSS can use to investigate.

I'm guessing at this based on the discussion in https://github.com/Azure/azure-sdk/issues/576 and the linked documentation that describes how ARM uses these values and flows them to downstream RPs.

But, @weikanglim is the expert here, so he can check my work.

Comment (vhvb1989) [reply cli/azd/cmd/up.go]: IIRC, the requestId is automatically generated by the Azure SDK Core pipeline.

Comment (ellismg) [reply cli/azd/cmd/up.go]: > IIRC, the requestId is automatically generated by the Azure SDK Core pipeline.

It generates a unique ID per request, which it sets as `x-ms-client-request-id`, the `x-ms-correlation-request-id` header can be used to provide an ID which is the same across multiple requests which represent a logical operation (it sort of feels like a `traceparent` in distributed tracing parlance).  I think this makes it easier for support folks to do queries like: "show me all the requests that happened across a single invocation of `azd provision`)

Comment (weikanglim) [reply cli/azd/cmd/up.go]: @ellismg You're indeed correct. We already have a `traceId` generated by default in our root span. We would use it to set  `x-ms-correlation-request-id`.  I updated #992

**Review by ellismg — APPROVED** [review]

File: cli/azd/cmd/down.go [hunk 9]
```diff
@@ -97,34 +98,38 @@ func newDownAction(
...
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/cmd/down.go:102]: I though `plan` was making sure to have the latest values updated in the environment... as someone might changed stuff somewhere else?  I don't know...  We need to investigate it deeper
